cql3/expr/expression.hh

```diff
@@ -97,6 +97,20 @@ using expression = std::variant<bool, conjunction, binary_operator, column_value
 template <typename T>
 concept ExpressionElement = utils::VariantElement<T, expression>;
 
+// An expression that doesn't contain subexpressions
+template <typename E>
+concept LeafExpression
```

Comment:
Optional: if we make this a real type, there's much less risk of it getting out-of-date.  At the price of making expression less flat, of course.

Reply:
This falls into the OO trap: there is one favored way to categorize expressions, and anything else has to use secondary means.

With these concepts, any subset of expression types can be made into a category. They can overlap.

Reply:
OK.  We'll just have to maintain it vigilantly.

Reply:
If we miss one, won't the overloaded_visitor code above complain during compilation, that some of the cases aren't covered?

Reply:
It will, but you could fix it by adding the branch explicitly to the visitor (instead of adding the type to LeafExpression).